api/controllers/apps.go

```diff
@@ -100,6 +100,23 @@ func AppDelete(rw http.ResponseWriter, r *http.Request) *httperr.Error {
 		return httperr.Errorf(404, "invalid app: %s", name)
 	}
 
+	// check to see if the app is linked to a service, if it is fail the request with precondition failed
+	services, err := models.ListServices()
+	if err != nil {
+		return httperr.Server(err)
+	}
+	for _, serviceName := range services {
+		service, err := provider.ServiceGet(serviceName.Name)
+		if err != nil {
+			return httperr.Server(err)
+		}
+		for _, linkedApp := range service.Apps {
+			if linkedApp.Name == app.Name {
+				return httperr.Errorf(412, "%s is linked to the service %s. You must delete all service links before deleting the app %s", app, service.Name, app)
```

Comment:
This will return on the first service it finds, correct? Does it make sense to do so, or to find the full list of linked services and then return a message enumerating all of them? Could save an annoying cycle of delete app, delete service, delete app, delete service.. but maybe it's not common to have too many linked services on the other hand.